scripts/valhalla_build_extract.py

```diff
@@ -0,0 +1,179 @@
+    """Actually creates the tar ball. Break out of main function for testability."""
+    if not tiles_fp_.exists() or not tiles_fp_.is_dir():
+        LOGGER.critical(f"Directory 'mjolnir.tile_dir': {tiles_fp_} was not found on the filesystem.")
+        sys.exit(1)
+
+    tiles_count = get_tile_count(tiles_fp_)
+    if not tiles_count:
+        LOGGER.critical(f"Directory {tiles_fp_} does not contain any usable graph tiles.")
+        sys.exit(1)
+
+    # write the in-memory index file
+    index_size = INDEX_BIN_SIZE * tiles_count
+    index_fd = BytesIO(b'0' * index_size)
+    index_fd.seek(0)
+
+    # first add the index file, then the tile dir to the tarfile
+    with tarfile.open(extract_fp_, 'w') as tar:
+        tarinfo = tarfile.TarInfo(INDEX_FILE)
+        tarinfo.size = index_size
+        tarinfo.mtime = int(time())
+        tar.addfile(tarinfo, index_fd)
+
+        tar.add(str(tiles_fp_), recursive=True, arcname='')
+
+    # get the offset and size from the tarred tile members
+    index: List[Tuple[int, int, int]] = list()
+    with tarfile.open(extract_fp_, 'r|') as tar:
+        for member in tar.getmembers():
+            if member.name.endswith('.gph'):
+                LOGGER.debug(f"Tile {member.name} with offset: {member.offset_data}, size: {member.size}")
+
+                index.append((member.offset_data, get_tile_id(member.name), member.size))
+
+    # write back the actual index info
+    with open(extract_fp_, 'r+b') as tar:
+        # jump to the data block, index.bin is the first file
+        tar.seek(BLOCKSIZE)
+        for entry in index:
+            tar.write(struct.pack(INDEX_BIN_FORMAT, *entry))
+
+    LOGGER.info(f"Finished tarring {tiles_count} tiles to {extract_fp_}")
+
+    # exit if no traffic extract wanted
+    if not do_traffic:
+        index_fd.close()
+        sys.exit(0)
+
+    LOGGER.info(f"Start creating traffic extract...")
+
+    # we already have the right size of the index file, simply reset it
+    index_fd.seek(0)
+    with tarfile.open(extract_fp_) as tar_in, tarfile.open(traffic_fp_, 'w') as tar_traffic:
+        # get a reference to the
+        in_fileobj = tar_in.fileobj
+
+        # add the index file as first data
+        tarinfo = tarfile.TarInfo('index.bin')
+        tarinfo.size = index_size
+        tarinfo.mtime = int(time())
+        tar_traffic.addfile(tarinfo, index_fd)
+        index_fd.close()
+
+        for tile_in in tar_in.getmembers():
+            if not tile_in.name.endswith('.gph'):
+                continue
+            # jump to the data's offset and skip the uninteresting bytes
+            in_fileobj.seek(tile_in.offset_data + GRAPHTILE_SKIP_BYTES)
+
+            # read the appropriate size of bytes from the tar into the TileHeader struct
+            tile_header = TileHeader()
+            b = BytesIO(in_fileobj.read(ctypes.sizeof(TileHeader)))
+            b.readinto(tile_header)
+            b.close()
```

Comment:
jumping straight to the one single `uint64_t` we're interested in and reading it into the TileHeader bit fielded struct